tdrs-backend/tdpservice/search_indexes/parsers/preparser.py

```diff
@@ -0,0 +1,169 @@
+            'state_fips':   {'type': 'string', 'required': True, 'regex': '^[0-9]{2}$'},
+            'tribe_code':   {'type': 'string', 'required': False, 'regex': '^([0-9]{3}|[ ]{3})$'},
+            'program_type': {'type': 'string', 'required': True, 'allowed': ['TAN', 'SSP']},
+            'edit':         {'type': 'string', 'required': True, 'allowed': ['1', '2']},
+            'encryption':   {'type': 'string', 'required': True, 'allowed': ['E', ' ']},
+            'update':       {'type': 'string', 'required': True, 'allowed': ['N', 'D', 'U']},
+        }
+
+        validator = Validator(header_schema)
+        is_valid = validator.validate(header)
+
+        return is_valid, validator
+
+    except Exception as e:
+        logger.error('Exception validating header row, please see error.')
+        logger.error(e)
+        return False, e
+    # should we close f?
+
+def validate_trailer(row, data_type, section):
+    """Validate the trailer row."""
+    """
+    https://www.acf.hhs.gov/sites/default/files/documents/ofa/transmission_file_header_trailer_record.pdf
+    length of 24
+    DESCRIPTION LENGTH FROM TO COMMENT
+    Title 7 1 7 Value = TRAILER
+    Number of Records 7 8 14 Right Adjusted
+    Blank 9 15 23 Value = spaces
+    Example:
+    'TRAILER0000001         '
+    """
+
+    logger.info('Validating trailer row.')
+    # Validate the trailer row
+    is_valid = True  # TODO: Implement validation logic with regex probably
+    errors = {}
+    return is_valid, errors
+
+
+def preparse(data_file_id, data_type, section):
+    """Validate metadata then dispatches file to appropriate parser."""
+    datafile = DataFile.objects.get(id=data_file_id).file
+    logger.debug("type %s", type(datafile))
+    logger.error("whatdo %s", dir(datafile))
+
+    # check file type and extension #TODO: this should be done by the frontend but let's verify here
+    # validate header and trailer lines
+    header_is_valid, header_errors = validate_header(datafile, data_type, section)
+    trailer_is_valid, trailer_errors = validate_trailer(datafile, data_type, section)
+    errors = header_errors  # + trailer_errors (how to combine Exception and dict? or logic around it)
+    if header_is_valid and trailer_is_valid:
+        logger.info("Preparsing succeeded.")
+    else:
+        logger.error("Preparse failed: %s", errors)
+        return
+        # return ParserLog.objects.create(
+        #    data_file=args.file,
+        #    errors=errors,
+        #    status=ParserLog.Status.REJECTED,
+        # )
+
+    # validate datatype and section
+
+    # given dict of data_type to parser function, call the correct one with arguments
+    if data_type == 'TANF':
+        tanf_parser.parse(datafile, section)
+    # elif data_type == 'SSP':
+    #    ssp_parser.switch(datafile, section)
+    # elif data_type == 'Tribal TANF':
+    #    tribal_tanf_parser.switch(datafile, section)
+    else:
+        raise Exception("Invalid data type.")
+
+
```

Comment:
I was working with this file directly initially so added this argparse block here. I think it's still fine, we might find some utility for it but could be convinced to delete it as well now that we have the unit-test harness for throwing files at this.